IC-1996: Use session-based drafts for multi page flows

lawrence-forooghian opened the pull request:
## What does this pull request do?

Proposes a consistent manner for building multi-page journeys where there isn't a permanent backing data store for the data submitted on each page.

## What is the intent behind these changes?

To allow us to build things like the check your answers page for the scheduling journey: https://dsdmoj.atlassian.net/browse/IC-1919, and to tidy up existing journeys like the referral assignment and cancellation.

## Next steps after merge

- test on dev
- &gt; 1 week after deploy, remove the legacy routes / actions (IC-2022)

lawrence-forooghian commented on doc/architecture/decisions/0011-use-session-based-drafts.md:
```diff
@@ -0,0 +1,76 @@
+
+## Status
+
+Accepted
```
(This is me being hopeful)

tomsmyers commented on server/services/draftsService.test.ts:
```diff
@@ -0,0 +1,174 @@
+        data: { a: 1, b: 2 },
+      })
+
+      expect(req.session['draft-2dc7ef56-58dd-4339-9924-c33318738068']).toEqual({
```
nit: i think it would be clearer to use a seperator that's not in the uuid here e.g. `draft:...` or `draft/...`

lawrence-forooghian replied:
Sure thing, I'll change it.

tomsmyers commented on server/services/draftsService.test.ts:
```diff
@@ -0,0 +1,174 @@
+
+        const draftsService = new DraftsService(clock)
+
+        expect(() => draftsService.fetchDraft('def456', req)).toThrow()
```
how often will calls want to do fetches to check existence? e.g. try { fetchDraft(...) } catch { // no draft, do something else }

if it's a pattern we expect, it's more efficient to return a value here e.g. `null` rather than throw an execption.

lawrence-forooghian replied:
I think that the "something else" will always be to show some sort of an error page. But, as I've mentioned in one of the commits, the error handling should probably be per-journey anyway, to show useful information about how to restart the journey. So probs makes sense to get rid of the generic error and handle on a case-by-case basis.

tomsmyers commented on server/services/draftsService.test.ts:
```diff
@@ -0,0 +1,174 @@
+
+        draftsService.deleteDraft('2dc7ef56-58dd-4339-9924-c33318738068', req)
+
+        expect(req.session['draft-2dc7ef56-58dd-4339-9924-c33318738068']).toBeUndefined()
```
i would probably expect this to behave like fetch and update, i.e. throw on missing draft

tomsmyers commented on server/services/draftsService.test.ts:
```diff
@@ -0,0 +1,174 @@
+
+        draftsService.deleteDraft('2dc7ef56-58dd-4339-9924-c33318738068', req)
+
+        expect(req.session['draft-2dc7ef56-58dd-4339-9924-c33318738068']).toBeUndefined()
```
nit: maybe better to check that `fetchDraft` throws here instead of testing the underlying implementation detail

lawrence-forooghian replied:
It kinda feels like that's then also testing `fetchDraft`, I dunno…

tomsmyers commented on server/services/draftsService.ts:
```diff
@@ -0,0 +1,74 @@
+import { Request } from 'express'
+import uuid from 'uuid'
+
+export interface Draft<Data> {
```
nit: i'd find this more readable with a conventional type param e.g. `T`

lawrence-forooghian replied:
👍

tomsmyers commented on server/services/draftsService.ts:
```diff
@@ -0,0 +1,74 @@
+  data: Data
+}
+
+export interface DraftDTO {
```
nit: do we need a DTO that is exactly the same as the underlying data type?

tomsmyers commented on server/services/draftsService.ts:
```diff
@@ -0,0 +1,74 @@
+export default class DraftsService {
+  constructor(private readonly clock: Clock) {}
+
+  private sessionKey(id: string): string {
```
does the UUID package provider a UUID type? could we use it here?

tomsmyers replied:
and else where in the below method signatures

lawrence-forooghian replied:
Doesn't look like it, just seems to be `string`.

tomsmyers commented on server/services/draftsService.ts:
```diff
@@ -0,0 +1,74 @@
+  data: Data
+}
+
+export interface DraftDTO {
```
does this need to be exported?

lawrence-forooghian replied:
Nope, good spot 👍

tomsmyers commented on server/services/draftsService.ts:
```diff
@@ -32,7 +33,11 @@ export default class DraftsService {
 
     if (dto === undefined) {
-      throw new Error(`Draft with ID ${id} not found in session`)
+      throw createError(404, new Error(`Draft with ID ${id} not found in session`), {
```
minor nit: you can just pass the string as the second arg

tomsmyers replied:
do we wanna be alerted to users hitting this error? if so, we should return 500 here, just because 404s are ignored in our monitoring tools

lawrence-forooghian replied:
Yeah, I think I'd like to see how often it happens for now, to see how much we need to invest in better error messaging. So let's make it a 500 to start

lawrence-forooghian commented:
@tomsmyers thanks for the review so far. I've addressed all your comments. Would you like me to split this up or are you happy to review the whole thing?

andrewburrowsmoj commented on server/services/draftsService.ts:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  private fetchDraftDTO(id: string, req: Request): DraftDTO | null {
+    const dto = req.session[this.sessionKey(id)]
```
I see that we're using the RedisStore solution for session store. So I think it would be worthwhile testing this in an environment with multiple containers just to be certain its working as expected.

Looking at the config, I believe that the session store time-out is currently 120 minutes. Is this an expectable time-out for saving drafts? I think the users may expect their drafts to last a bit longer. Perhaps some wording to inform the users that their drafts will be saved for x minutes might be useful?

andrewburrowsmoj replied:
Oh Sorry, I see that you made a comment about this in your documentation. 

You mention that it is 16 hours. How did you derive this?

From the code in `app.ts` I see: 

```
      cookie: { secure: config.https, sameSite: 'lax', maxAge: config.session.expiryMinutes * 60 * 1000 },
```

And the expiration defined in `config.ts`:

```
    expiryMinutes: Number(get('WEB_SESSION_TIMEOUT_IN_MINUTES', '120')),
```

I can't see `WEB_SESSION_TIMEOUT_IN_MINUTES` being set within the helm scripts so assuming it falls back to 120 minutes.

andrewburrowsmoj replied:
Oh Sorry... I see that you made a comment about this in your documentation. 

I was looking at the timeout for the cookie rather than timeout for the session in redis store. You mention that it is 16 hours just curious where this is defined?

lawrence-forooghian replied:
You know what, now that you ask, I have no idea how I managed to arrive at 16 hours. The line that I've linked to in `app.ts` in the documentation clearly shows it to be 120 minutes, like you say. I'll update the documentation to reflect this.

As to whether it's an acceptable timeout, I'm not sure. It probably comes down to the length of the journeys that we're using this for.

An alternative would be to, instead of storing these drafts in the session, to store them centrally in Redis, keyed by the username of the logged-in user. That way they wouldn't expire with the session. We could set a long TTL on these, perhaps a few days, after which Redis would clean them up. Do you think that might be a better idea? I can make that change if so.

lawrence-forooghian replied:
> So I think it would be worthwhile testing this in an environment with multiple containers just to be certain its working as expected before deploying to production? Dev environment has 2 containers.

Yep, good idea. I was wondering if you know if any way to check which container a request gets routed to? (So that I can check it really is working with 2 containers and that it's not just that all the requests I test it with are coincidentally being routed to the same container.)

andrewburrowsmoj replied:
> I was wondering if you know of any way to check which container a request gets routed to? 

We can simple do `kubectl log -f` on both containers. 
I suppose we can also test it locally by deleting the intervention-ui container and then starting it back up between requests.

andrewburrowsmoj replied:
> An alternative would be to, instead of storing these drafts in the session, store them directly in Redis, keyed by the username of the logged-in user. That way they wouldn't expire with the session. 

My understanding is that the express session being backed with the Redis store will automatically store them directly in Redis straight away (otherwise we'd have the problem with multiple containers and may need to play around with the config)....

andrewburrowsmoj replied:
I have another question (maybe for David) - when we redeploy, does the Redis container bounce as well? In which case all the session store will be lost given it's in-memory store.

andrewburrowsmoj replied:
> We could set a long TTL on these, perhaps a few days, after which Redis would clean them up. Do you think that might be a better idea? I can make that change if so.

hmmm, I don't think it would be an issues setting it for a few days...

However, in order to reduce problems... we could inform the user that their draft will only be saved for x amount of minutes/hours. I don't know how to present that to the user though...

andrewburrowsmoj replied:
Sorry to be polluting this comments ... another possible downside is that their session will not persist between browsers or different locations. As I think that the session store is based on the cookies stored in the browser.

lawrence-forooghian commented:
A few notes from what @andrewburrowsmoj and I just talked about:

- if the session were being wiped on deploy that'd mean everyone was being logged out on deploy, which I don't think is the case
- perhaps 2 hours is enough…
- but it might be nice to have the flexibility to set this lifetime independent of that of the session (increasing session lifetime has implications we don't necessarily want, like keeping user logged in longer), and would also mean it worked in multiple browsers (not really an aim, but nice if we get it for free)
- so I'll investigate putting stuff directly into Redis
- I'll confirm that updates to Redis are done synchronously before the response is returned (to avoid multiple container issues)
- speak to David to find out how to get `kubectl` working

lawrence-forooghian commented:
Superseded by #651.
